java/core/src/main/java/com/microsoft/bond/Bonded.java

```diff
@@ -5,106 +5,124 @@
         ArgumentHelper.ensureNotNull(bondType, "bondType");
-
-        // TODO: return stream-backed implementation with type constraint
-        throw new UnsupportedOperationException();
+        return new SpecificTypeObjectBonded<T>(objectInstance, bondType);
     }
 
     /**
-     * Creates a new Bonded instance from an input stream that is unconstrained by any struct types.
+     * Creates a new Bonded instance from an instance of a struct type, with the underlying Bond type obtained
+     * from the instance by calling the {@link BondSerializable#getBondType()} method.
      *
-     * @param inputStream input stream containing the serialized representation of this Bonded value
-     * @return new Bonded instance backed by the stream
+     * @param objectInstance the object backing this Bonded
+     * @param <T>            the struct type
+     * @return new Bonded instance backed by the object
      * @throws IllegalArgumentException if an argument is null
      */
-    public static Bonded<?> fromStream(InputStream inputStream) {
-        ArgumentHelper.ensureNotNull(inputStream, "inputStream");
-
-        // TODO: return stream-backed implementation without type constraint
-        throw new UnsupportedOperationException();
+    public static <T extends BondSerializable> Bonded<? extends T> fromObject(T objectInstance) {
+        ArgumentHelper.ensureNotNull(objectInstance, "objectInstance");
+        return new WildcardTypeObjectBonded<T>(objectInstance);
     }
 
     /**
-     * Creates a new Bonded instance from an instance of a struct type.
+     * Returns the default type descriptor associated with this Bonded instance or null if unspecified.
+     * The type descriptor determines the behavior of this Bonded instance,
+     * such as allowed conversion rules and default serialization/deserialization.
      *
-     * @param objectInstance
-     * @param bondType       the type descriptor of the struct type
-     * @param <T>            the struct type
-     * @return new Bonded instance backed by the object
-     * @throws IllegalArgumentException if an argument is null
+     * @return the Bond type descriptor or null if unspecified.
      */
-    public static <T extends BondSerializable> Bonded<T> fromObject(
-            T objectInstance, StructBondType<T> bondType) {
-        ArgumentHelper.ensureNotNull(objectInstance, "objectInstance");
-        ArgumentHelper.ensureNotNull(bondType, "bondType");
-
-        // TODO: return object-backed implementation
-        throw new UnsupportedOperationException();
-    }
+    public abstract StructBondType<? extends T> getBondType();
 
     /**
-     * Serializes content of the Bonded instance to a protocol writer.
+     * Serializes content of this Bonded instance to a protocol writer using the default Bond type descriptor.
+     * Throws an exception if there is no default type descriptor associated with this Bonded instance.
```

Comment:
Correct. The non-generic serialize/deserialize methods rely on the default type and throw when it's not specified.

This reminds me to add a similar comment to deserialize as well. Thanks.

---
In reply to: [130734564](https://github.com/Microsoft/bond/pull/540#discussion_r130734564) [](ancestors = 130734564)